packages/advanced-logic/specs/payment-context-exchange-rate.md

```diff
@@ -0,0 +1,122 @@
+
+## Description
+
+This extension describes how a request denominated in one currency should be paid in another currency.
+
+When this extension exists, its currency overrides the request currency from a payment point of view.
+The request denomination currency is used to compute the amount in another currency at the payment.
+
+The typical scenario is a request denominated in fiat and paid in ERC20.
+
+Many Exchange rates extensions can be associated with one request, together they describe:
+* To the payment processor what are the possible payment currencies and how to compute corresponding amounts
+* To the payment detection how to compare payments in one currency with a balance in another currency
+
+The version 0.1.0 is not designed to handle properly multiple payments, although it will detect them and
+compute a balance that is very close to the one payers and payee expect.
+
+## Properties
+
+| Property             | Type     | Description                         | Requirement   |
+|----------------------|----------|-------------------------------------|---------------|
+| **id**               | String   | constant value: "pc-exchange-rate"  | **Mandatory** |
+| **type**             | String   | constant value: "paymentContext"    | **Mandatory** |
+| **version**          | String   | constant value: "0.1.0"             | **Mandatory** |
+| **values**           | Object   |                                     |               |
+| **values.oracle**    | String   | TODO:should be fixed for the moment | **Mandatory** |
+| **values.timeframe** | Integer  | Exchange rate timespan              | **Mandatory** |
+| **values.currency**  | Currency | Currency of the expected amount     | **Mandatory** |
+
+---
+
+## Interpretation
+
+When this extension is present at least once, the request payment should be made in one of the exchange rate
+extensions currencies.
+
+### To initiate a payment
+
+The first step for payment initation is to pick one `values.currency` and its related exchange rate extension.
+There should be a payment network for this currency bound to the request.
+
+For the selected currency, the payment processor should look at the exchange rate given by the `values.oracle`.
+Once the exchange rate is fetched, the payment processor has a maximum `values.timeframe`to execute the payment, 
+in seconds. The timeframe will be computed based on the payment transaction datetime, so the payment processor 
+should anticipate network delays by adding a time margin.
+
+### To detect payments
+
+#### Standard case: one payment
+
+The first step is to detect the payment as described in the payment network extensions.
+
+If one payment is detected, we can compute an observed exchange rate, that we can compare to the oracle rate.
+The observed rate should be between the minimum and maximum oracle exchange rate of the period:
+`[paymentDateTime - values.timeframe, paymentDateTime]`.
+
+#### Edge case: many payments
```

Comment:
I don't see the issue with this one.
I would say that for each payment, we need to get the payment time and apply the rate at that time, then sum everything to get the balance.

Reply:
The issue is that binding a payment rate to a payment time creates a different between what the payer wants to pay in fiat and what he actually pays. Basically because of the time it takes for his transaction to be mined. 
An ideal solution would be to apply the same mechanism to every sub-payment. Ideal = for later.

Reply:
I don't see how it differs from the mechanism for one payment. If it's the same mechanism, but done for every payment, that would actually be simpler, since we already iterate over all the payments, and we wouldn't need a special case.